From 83b20d547022a6dd8ffe1902a25f93ae5cfc1765 Mon Sep 17 00:00:00 2001
From: "Gadiraju, Ashokkumar" <agadiraju@irobot.com>
Date: Thu, 27 Jun 2019 09:09:13 -0400
Subject: [PATCH 1/2] sem: set appropriate errno and return value in sem_close

Adding appropriate errno and return value in sem_close for semaphore
validity/found and unmap result to caller for better error handling
---
 src/thread/sem_open.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/thread/sem_open.c b/src/thread/sem_open.c
index de8555c5..e731227e 100644
--- a/src/thread/sem_open.c
+++ b/src/thread/sem_open.c
@@ -163,11 +163,17 @@ int sem_close(sem_t *sem)
 	int i;
 	LOCK(lock);
 	for (i=0; i<SEM_NSEMS_MAX && semtab[i].sem != sem; i++);
+	if (i >= SEM_NSEMS_MAX) {
+		errno = ENOENT;
+		UNLOCK(lock);
+		return -1;
+	}
 	if (!--semtab[i].refcnt) {
 		semtab[i].sem = 0;
 		semtab[i].ino = 0;
 	}
 	UNLOCK(lock);
-	munmap(sem, sizeof *sem);
+	errno = -munmap(sem, sizeof *sem);
+	if (errno) return -1;
 	return 0;
 }
-- 
2.22.0.214.g8dca754b1e

